### DISCUSSION TITLE
Open-source Lombiq projects now published on NuGet

### DISCUSSION THREAD

**Discussion by Piedone** [OrchardCMS/OrchardCore#11294] [Show and tell]
Most of the Lombiq modules and libraries are available on NuGet now here: https://www.nuget.org/profiles/Lombiq So, you can just install and use them right away.

Do let me know if you can figure out how it's possible to package up projects that you also need to restore as NPM packages :). See the issue here: https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/issues/48

**Comment by Skrypt** [comment]
We need to start using **Webpack** at some point and that's why there's not been much progress on this subject lately; Webpack would allow doing an `npm publish`. I think, if a module contains a javascript library that should be published on NPM then that module should not contain the JS source code but just a reference to the resulting NPM package(s) assets. If we do include some source code javascript assets in our solutions they should be added in the /Assets folder of a module because this specific folder is never embedding files when the module is built/published. 

Then, the remaining is about the fact that we should automate the building of the javascript assets from the CI. This can be done with either Gulp or Webpack easily.

So basically, just before doing the `dotnet build` we should run a `npm install` and then a `gulp build` which would produce the proper .css and .js files in the wwwroot folders for them to be embedded when we run the `dotnet publish` for example.

That would be an improvement for "reproducible builds" because we would be forced for example to document which version of Node.js + NPM is required to build a specific version of OC and this would be part of the source code at one point in time.

**Reply by Piedone** [reply]
The problem under https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/issues/48 is that something like the mentioned Node.js Extensions project, if you want to use it from your own projects, it needs to restore NPM packages locally (from a packages.json), i.e. it needs to be included either as source or as an NPM package. However, it also needs to include targets for MSBuild, and thus needs to be a NuGet package as well. I don't yet know how we can bring the two together, apart from actually publishing both an NPM and NuGet package.

BTW note that Gulp yet [doesn't officially support the previous Node LTS release, 14.x](https://github.com/gulpjs/gulp/discussions/2649#discussioncomment-1880077), nor anything after that. We try to get away from task runners by just using Node.js scripts in the [Node.js Extensions](https://github.com/Lombiq/NodeJs-Extensions/) project, and it looks good so far, and one fewer dependency.

**Reply by Skrypt** [reply]
Yes, we would need to have an npm repository for Orchard Core so that we can provide a common set of tools for Orchard Core. This way we could manage independently the dependencies and a project like OrchardCore.Resources could embed these NPM packages. I have no control over this as this is a decision that needs to be made conjointly with @sebastienros because, in the end, he needs to have a view of this NPM repository. I remember that @jptissot needed to make an NPM repository too for the functional tests back then. I don't remember exactly what he did in the end.

But, without analyzing too much in detail what you need on your side, I believe that we reach a point of no return with the gulp pipeline. It will need a refresh either by using your method or WebPack which seems quite popular; probably because it has more feature.

**Reply by Piedone** [reply]
Good points, thank you!

**Comment by Piedone** [comment]
With the [Lombiq Node.js Extensions](https://github.com/Lombiq/NodeJs-Extensions) project we're getting there to offer Node builds from NuGet. Specifically, SCSS build works now and soon will have the rest of what [Gulp Extensions](https://github.com/Lombiq/Gulp-Extensions) does.

**Comment by 0liver** [comment]
[Lombiq Node.js Extensions](https://github.com/Lombiq/NodeJs-Extensions) has finally been publicly released. It can be used via a project reference (using submodules), or from its [NuGet package](https://www.nuget.org/packages/Lombiq.NodeJs.Extensions/).

Highlights:
- Hooks into MSBuild's build process: reports warnings and errors as regular MSBuild warnings and errors with clickable links to the offending line of code
- SCSS linting and compilation pre-configured, no additional config needed
- JS linting and compilation pre-configured, no additional config needed
- Markdown linting against some sensible defaults, opt-in
- Arbitrary asset copying from any source path to any target path, e.g. _wwwroot_, needs to be configured
- Simple configuration from _package.json_
- All processed assets will be embedded in the project's assembly as if _wwwroot_ had been there from the beginning
- Usable from a NuGet package

Lombiq Node.js Extensions leverages Node.js' `corepack` tool to use `pnpm` for NPM package management and script execution. That's why Node.js >=14.19.0 or Node.js >=16.9.0 are mandatory.
